libs/filamentjs/utilities.js

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ---------------
+// Buffer Wrappers
+// ---------------
+
+// These wrappers make it easy for JavaScript clients to pass large swaths of data to Filament. They
+// copy the contents of the given typed array into the WASM heap, then return a low-level buffer
+// descriptor object. If the given array was taken from the WASM heap, then they create a temporary
+// copy because the input pointer becomes invalidated after allocating heap memory for the buffer
+// descriptor.
+
+Filament.Buffer = function(typedarray) {
+  console.assert(typedarray.buffer instanceof ArrayBuffer);
+  console.assert(typedarray.byteLength > 0);
+  if (Filament.HEAPU32.buffer == typedarray.buffer) {
+      typedarray = new Uint8Array(typedarray);
```

Comment:
You have inconsistent indents in this file: 2 vs 4 spaces (or are there tabs?)

Reply:
Good catch, fixed.